parquet-variant/src/main/java/org/apache/parquet/variant/VariantElementConverter.java

```diff
@@ -0,0 +1,668 @@
+import org.apache.parquet.schema.PrimitiveType;
+import org.apache.parquet.schema.Type;
+
+/**
+ * Converter for a shredded Variant containing a value and/or typed_value field: either a top-level
+ * Variant column, or a nested array element or object field. The top-level converter is handled
+ * by a subclass (VariantColumnConverter) that also reads metadata.
+ *
+ * Converters for the `value` and `typed_value` fields are implemented as nested classes in this
+ * class.
+ *
+ * All converters for a Variant column append their results to a VariantBuilder as values are read from Parquet.
+ *
+ * Values in `typed_value` are appended by the child converter. Values in `value` are stored by a
+ * child converter, but only appended when completing this group. Additionally, object fields are
+ * appended by the `typed_value` converter, but because residual values are stored in `value`, this
+ * converter is responsible for finalizing the object.
+ */
+class VariantElementConverter extends GroupConverter implements VariantConverter {
+
+  // We need to remember the start position in order to tell if the child typed_value was non-null.
+  // Maybe we should just add a callback to each child's end()?
+  private int startWritePos;
+  private boolean typedValueIsObject = false;
+  private int valueIdx = -1;
+  private int typedValueIdx = -1;
+  protected VariantBuilderHolder holder;
+  protected Converter[] converters;
+
+  // The following are only used if this is an object field.
+  private String objectFieldName = null;
+  private int objectFieldId = -1;
+  private VariantObjectConverter parent = null;
+
+  // Only used if typedValueIsObject is true.
+  private Set<String> shreddedObjectKeys;
+
+  @Override
+  public void init(VariantBuilderHolder holder) {
+    this.holder = holder;
+    for (Converter converter : converters) {
+      if (converter != null) {
+        ((VariantConverter) converter).init(holder);
+      }
+    }
+  }
+
+  public VariantElementConverter(GroupType variantSchema, String objectFieldName, VariantObjectConverter parent) {
+    this(variantSchema);
+    this.objectFieldName = objectFieldName;
+    this.parent = parent;
+  }
+
+  public VariantElementConverter(GroupType variantSchema) {
+    converters = new Converter[variantSchema.getFieldCount()];
+
+    List<Type> fields = variantSchema.getFields();
+
+    for (int i = 0; i < fields.size(); i++) {
+      Type field = fields.get(i);
+      String fieldName = field.getName();
+      if (fieldName.equals("value")) {
+        this.valueIdx = i;
+        if (!field.isPrimitive() || field.asPrimitiveType().getPrimitiveTypeName() != BINARY) {
+          throw new IllegalArgumentException("Value must be a binary value");
+        }
+      } else if (fieldName.equals("typed_value")) {
+        this.typedValueIdx = i;
+      }
+    }
+
+    if (valueIdx >= 0) {
+      converters[valueIdx] = new VariantValueConverter(this);
+    }
+
+    if (typedValueIdx >= 0) {
```

Comment:
I think that this class is doing too much. I would rather have a separate class for each shredded case: primitive, object, and array. That would make the code easier to read and would avoid cases where you're keeping important state in unexpected places, like the branches in `end` that handle object or array based on whether `objectFieldName` is null.